### PR TITLE
Retrieve Books By Genre and Top Sellers

### PR DESCRIPTION
**Pull request by notrioht**
### Summary

We've elevated our book management system by introducing two pivotal `GET` endpoints. The first allows users to retrieve a curated list of books based on their genre, streamlining the search process and catering to specific reading preferences. The second showcases our top 10 best-selling books, offering a quick insight into the titles that have resonated most with our audience based on copies sold. These enhancements not only optimize the browsing experience by providing tailored book recommendations but also highlight trending literature, ensuring our readers always have access to the best of our catalog.

### Changes

- Book: Add a new instance variable for rating.
- BookController: Implemented the `GET` endpoint to retrieve books by genre and top-selling books.
- BookService: Added two service methods to retrieve books by genre and top-selling books.
- BookRepository: Update the necessary query for fetching books by genre and top-selling books.

### Endpoint Details
- Purpose: Allows users or administrators to retrieve a list of books by genre and top-selling books.
- Endpoint 1: `/api/books/genre/{genre}`
- Endpoint 2: `/api/books/top-sellers`
- HTTP Method: `GET`
- Response 1: Returns a JSON array containing detailed information about all books by genre in the system. Each book entry includes essential details such as the book's ISBN, title, description, price, author, genre, publisher, publication year, the number of copies sold, and rating.
- Response 2: Returns a JSON array containing detailed information about the top 10 selling books in the system. Each book entry includes essential details such as the book's ISBN, title, description, price, author, genre, publisher, publication year, the number of copies sold, and rating.
    Example Response:
```json
[
 {
  "id":"948e545e-b7bb-4594-8f8f-52259167e416",
  "isbn":"9780439064873",
  "name":"Harry Potter and the Chamber of Secrets",
  "description":"The second book in the Harry Potter series.",
  "price":20.99,"authorId":"c88664b2-2adb-4d1a-bc6c-0a2916be3e8c",
  "genre":"Fantasy",
  "publisher":"Scholastic",
  "yearPublished":1998,
  "copiesSold":77000000,
  "rating":0.0
 }
]
```

Test Cases
- Retrieve the list of all books by genre / by top-seller.
- Check for an empty list if there are no books in the system.
- Verify the response structure and data accuracy.

How to Test
- Pull the branch to your local setup.
- Run the application.
- Use the provided curl command or any API tool like Postman to test the endpoint.

```cmd
curl http://localhost:8080/api/books/genre/fantasy
```

### CODE REVIEW

**Comment by aaronthangnguyen** [comment]
Approved with a few suggestions:
1. Write tests for both endpoints
2. Run `make fmt-apply` to format the project
3. Update `src/main/resources/dummy.sql`

File: src/main/java/com/team23/geektext/book/BookController.java [hunk 1]
```diff
@@ -61,4 +60,19 @@ public ResponseEntity<?> getBookByIsbn(@PathVariable String isbn) {
...
+        List<Book> books = bookService.getAllBooksByGenre(genre);
+        return new ResponseEntity<List<Book>>(books, HttpStatus.OK);
+    }
+
+    @GetMapping("/top-seller")
+    public ResponseEntity<List<Book>> findTop10ByOrderByCopiesSoldDesc() {
+        List<Book> books = bookService.findTop10ByOrderByCopiesSoldDesc();
+        return new ResponseEntity<List<Book>>(books, HttpStatus.OK);
+    }
+
```
Comment (aatng-gh) [inline src/main/java/com/team23/geektext/book/BookController.java:63]: Reformat with `make fmt-apply`

**Review by aatng-gh — APPROVED** [review]
@notrioht great job with this pull request!
Approved with a few suggestions:
1. Write tests for both endpoints
2. Run `make fmt-apply` to format the project
3. Update `src/main/resources/dummy.sql` with the new `rating` attribute

**Review by aatng-gh — CHANGES_REQUESTED** [review]
CleanShot 2023-10-05 at 14 34 13@2x
According to "Book Rating and Commenting" feature, rating shouldn't be Book's attribute. Rating is a class by itself. Apparently, this endpoint is dependent on @murphybrenna feature.

**Review by notrioht — COMMENTED** [review]
Reviewed

**Review by notrioht — COMMENTED** [review]
Checked

File: src/main/java/com/team23/geektext/book/Book.java [hunk 2]
```diff
@@ -28,6 +27,8 @@ public class Book {
     private int yearPublished;
     private int copiesSold;
 
+    private double rating;
```
Comment (notrioht) [inline src/main/java/com/team23/geektext/book/Book.java]: done

**Review by notrioht — COMMENTED** [review]
done

**Pull request closed without merging**